src/v/storage/spill_key_index.cc

```diff
@@ -110,6 +130,7 @@ ss::future<> spill_key_index::add_key(compaction_key b, value_type v) {
                 node.mapped(),
                 [this](const bytes& k, value_type o) {
                     _keys_mem_usage -= k.size();
+                    _mem_units.return_units(k.size());
```

Comment:
the concern here is that return_units will throw if there is an accounting imbalance and the return units goes over the set amount of units available in the semaphore. maybe all the sizes are accounted for, but it's not clear that that is true with the size here coming out of the index. one option would be to look at available units in the semaphore and clamp if it is a concern.